File: backend/services/cleanup.js
var cache = require("../cache");
var config = require("../config");
var db = require("../lib/db");
var emitters = require("./emitters");
var visibility = require("./visibility");
var sos = require("./sos");

var log = config.log;

// io reference, set via setIo()
var _io = null;
function setIo(io) { _io = io; }

function startAll() {
    // ── Expire offline users ─────────────────────────────────────────────
    setInterval(function() {
        var now = Date.now();
        for (var [uid, entry] of cache.offlineUsers.entries()) {
            if (entry.expiresAt && entry.expiresAt <= now) {
                cache.offlineUsers.delete(uid);
                for (var [sid, u] of cache.activeUsers) {
                    if (visibility.canSee(u.userId, entry.user.userId)) _io.to(sid).emit("userDisconnect", entry.user.socketId);
                }
            }
        }
    }, 60 * 1000);

    // ── Expire watch tokens ──────────────────────────────────────────────
    setInterval(function() {
        var now = Date.now();
        for (var [token, entry] of cache.watchTokens.entries()) {
            if (entry.exp < now) {
                _io.to("watch:" + token).emit("watchUpdate", { user: null, sos: { active: false } });
                cache.watchTokens.delete(token);
            }
        }
    }, 30 * 1000);

    // ── Expire live tokens ───────────────────────────────────────────────
    setInterval(function() {
        var now = Date.now();
        for (var [token, entry] of cache.liveTokens.entries()) {
            if (entry.expiresAt && entry.expiresAt <= now) {
                cache.liveTokens.delete(token);
                // Maintain liveTokensByUser index
                var userSet = cache.liveTokensByUser.get(entry.userId);
                if (userSet) { userSet.delete(token); if (userSet.size === 0) cache.liveTokensByUser.delete(entry.userId); }
                _io.to("live:" + token).emit("liveExpired", { message: "Link expired" });
            }
        }
        db.deleteExpiredLiveTokens().catch(function(e) { log.error({ err: e.message }, "Failed to clean expired live tokens from DB"); });
    }, 60 * 1000);

    // ── Clean empty old rooms ────────────────────────────────────────────
    setInterval(function() {
        var now = Date.now(); var sevenDays = 7 * 24 * 60 * 60 * 1000;
        for (var [code, room] of cache.rooms.entries()) {
            if (room.members.size === 0 && now - room.createdAt > sevenDays) { cache.rooms.delete(code); }
        }
        db.deleteEmptyOldRooms(sevenDays).catch(function(e) { log.error({ err: e.message }, "Failed to clean old empty rooms from DB"); });
    }, 60 * 60 * 1000);

    // ── Expire time-limited room admins ─────────────────────────────────
    setInterval(function() {
        var now = Date.now();
        // In-memory expiry
        for (var [code, roleMap] of cache.roomMemberRoles) {
            for (var [uid, entry] of roleMap) {
                if (entry.role === "admin" && entry.expiresAt && entry.expiresAt <= now) {
                    entry.role = "member";
                    entry.expiresAt = null;
                    var room = cache.rooms.get(code);
                    if (room) {
                        for (var mid of room.members) {
                            var ms = emitters.findSocketByUserId(mid);
                            if (ms) ms.emit("roomAdminUpdated", { roomCode: code, userId: uid, role: "member", expiresAt: null });
                        }
                    }
                }
            }
        }
        db.expireRoomAdmins(now).catch(function(e) { log.error({ err: e.message }, "Failed to expire room admins in DB"); });
    }, 60 * 1000);

    // ── Expire time-limited guardianships ─────────────────────────────────
    setInterval(function() {
        var now = Date.now();
        for (var [gId, wardMap] of cache.guardianships) {
            for (var [wId, entry] of wardMap) {
                if (entry.status === "active" && entry.expiresAt && entry.expiresAt <= now) {
                    wardMap.delete(wId);
                    var updatePayload = { guardianId: gId, wardId: wId, status: "expired", expiresAt: null };
                    var gs = emitters.findSocketByUserId(gId);
                    if (gs) gs.emit("guardianUpdated", updatePayload);
                    var ws = emitters.findSocketByUserId(wId);
                    if (ws) ws.emit("guardianUpdated", updatePayload);
                }
            }
        }
        db.expireGuardianships(now).catch(function(e) { log.error({ err: e.message }, "Failed to expire guardianships in DB"); });
    }, 60 * 1000);

    // ── Check-in overdue polling ─────────────────────────────────────────
    setInterval(function() {
        var now = Date.now();
        var permissions = require("./permissions");
        // Pre-build manageable map once per tick: managerUserId → Set<wardUserId>
        // This avoids calling canManage() for every (user, user) pair (O(n^2) → O(n))
        var manageableMap = new Map();
        for (var u of cache.activeUsers.values()) {
            manageableMap.set(u.userId, permissions.getManageableUsers(u.userId));
        }
        for (var user of cache.activeUsers.values()) {
            if (!user.checkIn) user.checkIn = { enabled: false, intervalMinutes: 5, overdueMinutes: 7, lastCheckInAt: now };
            if (!user.checkIn.lastCheckInAt) user.checkIn.lastCheckInAt = now;
            if (!user.checkIn.enabled) continue;
            var intervalMs = user.checkIn.intervalMinutes * 60 * 1000;
            var overdueMs = user.checkIn.overdueMinutes * 60 * 1000;
            var since = now - user.checkIn.lastCheckInAt;
            if (since >= intervalMs) {
                _io.to(user.socketId).emit("checkInRequest", { intervalMinutes: user.checkIn.intervalMinutes, overdueMinutes: user.checkIn.overdueMinutes });
            }
            if (since >= overdueMs) {
                var missedPayload = { socketId: user.socketId, userId: user.userId, displayName: user.displayName, lastCheckInAt: user.checkIn.lastCheckInAt, overdueMinutes: user.checkIn.overdueMinutes };
                for (var mgr of cache.activeUsers.values()) {
                    if (mgr.userId !== user.userId) {
                        var mgrSet = manageableMap.get(mgr.userId);
                        if (mgrSet && mgrSet.has(user.userId)) {
                            _io.to(mgr.socketId).emit("checkInMissed", missedPayload);
                        }
                    }
                }
                sos.emitLiveCheckIn(user);
            }
        }
    }, 60 * 1000);
}

module.exports = {
    setIo: setIo,
    startAll: startAll
};

File: backend/internal/ws/cleanup.go
package ws

import (
	"context"
	"log/slog"
	"time"

	"kinnect-v3/internal/cache"
	"kinnect-v3/internal/db"
)

const (
	sevenDaysMs = 7 * 24 * 60 * 60 * 1000
)

// StartCleanupRoutines starts 7 periodic goroutines for cache and DB cleanup.
func (h *Hub) StartCleanupRoutines(ctx context.Context) {
	// 1. Expire offline users (every 60s)
	go func() {
		ticker := time.NewTicker(60 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				h.cleanupExpireOfflineUsers()
			}
		}
	}()

	// 2. Expire watch tokens (every 30s)
	go func() {
		ticker := time.NewTicker(30 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				h.cleanupExpireWatchTokens()
			}
		}
	}()

	// 3. Expire live tokens (every 60s)
	go func() {
		ticker := time.NewTicker(60 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				h.cleanupExpireLiveTokens()
			}
		}
	}()

	// 4. Clean empty old rooms (every 1h)
	go func() {
		ticker := time.NewTicker(1 * time.Hour)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				h.cleanupEmptyOldRooms()
			}
		}
	}()

	// 5. Expire time-limited room admins (every 60s)
	go func() {
		ticker := time.NewTicker(60 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				h.cleanupExpireRoomAdmins()
			}
		}
	}()

	// 6. Expire time-limited guardianships (every 60s)
	go func() {
		ticker := time.NewTicker(60 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				h.cleanupExpireGuardianships()
			}
		}
	}()

	// 7. Check-in overdue polling (every 60s)
	go func() {
		ticker := time.NewTicker(60 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				h.cleanupCheckInOverdue()
			}
		}
	}()

	slog.Info("Cleanup routines started")
}

func (h *Hub) cleanupExpireOfflineUsers() {
	now := time.Now().UnixMilli()
	list := h.cache.CollectExpiredOfflineUsers(now)
	for _, e := range list {
		h.SendToClients(e.VisibleSids, "userDisconnect", e.SocketID)
	}
}

func (h *Hub) cleanupExpireWatchTokens() {
	now := time.Now().UnixMilli()
	tokens := h.cache.CollectExpiredWatchTokens(now)
	payload := map[string]interface{}{"user": nil, "sos": map[string]interface{}{"active": false}}
	for _, token := range tokens {
		h.SendToGroup("watch:"+token, "watchUpdate", payload)
	}
}

func (h *Hub) cleanupExpireLiveTokens() {
	now := time.Now().UnixMilli()
	list := h.cache.CollectExpiredLiveTokens(now)
	for _, e := range list {
		h.SendToGroup("live:"+e.Token, "liveExpired", map[string]interface{}{"message": "Link expired"})
	}
	if len(list) > 0 {
		if _, err := db.DeleteExpiredLiveTokens(context.Background(), h.pool.DB); err != nil {
			slog.Error("Failed to delete expired live tokens from DB", "error", err)
		}
	}
}

func (h *Hub) cleanupEmptyOldRooms() {
	now := time.Now().UnixMilli()
	_ = h.cache.CollectEmptyOldRooms(now, sevenDaysMs)
	if err := db.DeleteEmptyOldRooms(context.Background(), h.pool.DB, sevenDaysMs); err != nil {
		slog.Error("Failed to delete empty old rooms from DB", "error", err)
	}
}

func (h *Hub) cleanupExpireRoomAdmins() {
	now := time.Now().UnixMilli()
	list := h.cache.ExpireRoomAdminsInCache(now)
	payload := func(roomCode, userID string) map[string]interface{} {
		return map[string]interface{}{
			"roomCode":   roomCode,
			"userId":     userID,
			"role":       "member",
			"expiresAt":  nil,
		}
	}
	for _, e := range list {
		p := payload(e.RoomCode, e.UserID)
		h.SendToClients(e.MemberSids, "roomAdminUpdated", p)
	}
	if len(list) > 0 {
		if _, err := db.ExpireRoomAdmins(context.Background(), h.pool.DB, now); err != nil {
			slog.Error("Failed to expire room admins in DB", "error", err)
		}
	}
}

func (h *Hub) cleanupExpireGuardianships() {
	now := time.Now().UnixMilli()
	list := h.cache.CollectExpiredGuardianships(now)
	updatePayload := func(gID, wID string) map[string]interface{} {
		return map[string]interface{}{
			"guardianId": gID,
			"wardId":     wID,
			"status":     "expired",
			"expiresAt":  nil,
		}
	}
	for _, e := range list {
		p := updatePayload(e.GuardianID, e.WardID)
		if c := h.GetClientByUserID(e.GuardianID); c != nil {
			c.Send("guardianUpdated", p)
		}
		if c := h.GetClientByUserID(e.WardID); c != nil {
			c.Send("guardianUpdated", p)
		}
	}
	if len(list) > 0 {
		if _, err := db.ExpireGuardianships(context.Background(), h.pool.DB, now); err != nil {
			slog.Error("Failed to expire guardianships in DB", "error", err)
		}
	}
}

func (h *Hub) cleanupCheckInOverdue() {
	now := time.Now().UnixMilli()
	h.cache.ForEachActiveUser(func(socketID string, user *cache.ActiveUser) {
		ch := &user.CheckIn
		if !ch.Enabled {
			return
		}
		lastAt := ch.LastCheckInAt
		if lastAt == 0 {
			lastAt = now
		}
		intervalMs := int64(ch.IntervalMin) * 60 * 1000
		overdueMs := int64(ch.OverdueMin) * 60 * 1000
		since := now - lastAt

		if since >= intervalMs {
			h.SendToClient(socketID, "checkInRequest", map[string]interface{}{
				"intervalMinutes":  ch.IntervalMin,
				"overdueMinutes":   ch.OverdueMin,
			})
		}
		if since >= overdueMs {
			missedPayload := map[string]interface{}{
				"socketId":        socketID,
				"userId":          user.UserID,
				"displayName":     user.DisplayName,
				"lastCheckInAt":   lastAt,
				"overdueMinutes":  ch.OverdueMin,
			}
			h.emitToVisible(user, "checkInMissed", missedPayload)
		}
	})
}
